File: CS50/intro/week2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Prompt user for text
    string text = get_string("Text: ");

    // Count number of letters, words, and sentences in text
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Calculate L; L is average number of letters per 100 words
    float L = (float) letters / words * 100;

    // Calculate S; S is average number of sentences per 100 words
    float S = (float) sentences / words * 100;

    // Compute the Coleman-Liau index
    float index = (float) 0.0588 * L - 0.296 * S - 15.8;

    // Print the grade level
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", (int) round(index));
    }
}

int count_letters(string text)
{
    // Keep track of letters
    int letters = 0;

    // Add count for each letter
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    return letters;
}

int count_words(string text) // Words end in spaces, punctuation can be ignored
{
    // Keep track of words
    int words = 0;

    // Add count when space
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isspace(text[i]))
        {
            words++;
        }
    }
    // Add one more for the end of passages because no space exists.
    words++;
    return words;
}

int count_sentences(string text) // Sentences end in ., !, or ?
{
    // Keep track of sentences
    int sentences = 0;

    // Add count when ., ?, or ! is found
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }
    return sentences;
}

File: CS50/intro/week6/readability.py
from cs50 import get_string


def main():
    # Prompt user for text
    text = get_string("Text: ")

    # Count number of ltrs, wrds, and stncs in text
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # Calculate L
    L = float(letters / words * 100)

    # Calculate S
    S = float(sentences / words * 100)

    # Compute C-L index
    index = float(0.0588 * L - 0.296 * S - 15.8)

    # Print grade level
    if (index < 1):
        print("Before Grade 1")
    elif (index > 16):
        print("Grade 16+")
    else:
        print("Grade", round(index))


def count_letters(text):
    letters = 0

    # Count each letter
    for i in range(len(text)):
        if text[i].isalpha():
            letters += 1

    return letters


def count_words(text):
    words = 0

    # Add count to words for every space
    for i in range(len(text)):
        if text[i].isspace():
            words += 1

    # Add one at end, because end has no space
    words += 1
    return words


def count_sentences(text):  # Sentences end in ., !, or ?
    sentences = 0

    # Add count when above are found
    for i in range(len(text)):
        if text[i] == "." or text[i] == "!" or text[i] == "?":
            sentences += 1

    return sentences


main()
